Update JVM Troubleshooting (#4688)
elastic/docs-content: troubleshoot/elasticsearch/high-jvm-memory-pressure.md

High JVM memory usage can degrade cluster performance and trigger [circuit breaker errors](circuit-breaker-errors.md). To prevent this, we recommend taking steps to reduce memory pressure if a node’s JVM memory usage consistently exceeds 85%.

## Diagnose high JVM memory pressure [diagnose-high-jvm-memory-pressure]

### Check JVM memory pressure [diagnose-check-pressure] 

> {{es}}'s JVM [uses a G1GC garbage collector](/deploy-manage/deploy/self-managed/bootstrap-checks.md). Over time this causes the JVM heap usage metric to reflect a sawtooth pattern as shown in [Understanding JVM heap memory](https://www.elastic.co/blog/jvm-essentials-for-elasticsearch). Due to this, we recommend focusing monitoring on the JVM memory pressure which gives a rolling average of old garbage collection and better represents the node's ongoing JVM responsiveness.

Reviewer: I think `rolling average` is more clear than `across time measurement`. otherwise lgtm

> You can also use the [nodes stats API](https://www.elastic.co/docs/api/doc/elasticsearch/operation/operation-nodes-stats) to calculate the current JVM memory pressure for each node.

Author: The sublinked ECE/ECH page is the platform does the calculation for you & shows it to you in a UI. So Also is to calculate it manually.

```console
GET _nodes/stats?filter_path=nodes.*.name,nodes.*.jvm.mem.pools.old
```

You can calculate the memory pressure from the ratio of `used_in_bytes` to `max_in_bytes`. For example, you can store this output into `nodes_stats.json` and then using [third-party tool JQ](https://jqlang.github.io/jq/):

```bash
cat nodes_stats.json | jq -rc '.nodes[]|.name as $n|.jvm.mem.pools.old|{name:$n, memory_pressure:(100*.used_in_bytes/.max_in_bytes|round) }'
```

> {{ech}} and {{ece}} also include a JVM memory pressure indicator for each node in your cluster in your deployment overview as discussed under their [memory pressure monitoring](/deploy-manage/monitor/ec-memory-pressure.md). These indicators turn red once JVM memory pressure reaches 75%.

Reviewer: I'm not finding the previous link

[timestamp_short_interval_from_last][INFO ][o.e.m.j.JvmGcMonitorService] [node_id] [gc][number] overhead, spent [21s] collecting in the last [40s]
```

> Garbage collection will also surface as part of the [nodes hot threads API](https://www.elastic.co/docs/api/doc/elasticsearch/operation/operation-nodes-hot-threads) output `OTHER_CPU` as described in [troubleshooting high CPU usage](/troubleshoot/elasticsearch/high-cpu-usage.md#check-hot-threads).

Reviewer 2: ```suggestion
Garbage collection activity can also appear in the output of the [nodes hot threads API](https://www.elastic.co/docs/api/doc/elasticsearch/operation/operation-nodes-hot-threads), under the `OTHER_CPU` category, as described in [troubleshooting high CPU usage](/troubleshoot/elasticsearch/high-cpu-usage.md#check-hot-threads).
```

If you have an [Elastic subscription](https://www.elastic.co/pricing), you can [request Elastic's assistance](/troubleshoot/index.md#contact-us) reviewing this output. When doing so, kindly:

* Grant written permission for Elastic to review your uploaded heap dumps within the support case.
* Share this file only after receiving any necessary business approvals as it might contain private information. Files are handled according to [Elastic's privacy statement](https://www.elastic.co/legal/privacy-statement).
* Share heap dumps through our secure [Support Portal](https://support.elastic.co/). If your files are too large to upload, you can request a secure URL in the support case.
* Share the [garbage collector logs](elasticsearch://reference/elasticsearch/jvm-settings.md#gc-logging) covering the same time period.

## Monitor JVM memory pressure [monitor-jvm-memory-pressure]

:::{include} /deploy-manage/_snippets/autoops-callout-with-ech.md
:::

To track JVM over time, we recommend enabling monitoring

:::::::{applies-switch}

::::::{applies-item} { ess:, ece: }

> * (Recommend) Enable [AutoOps](/deploy-manage/monitor/autoops.md)

Reviewer: ```suggestion
* (Recommend) Enable [AutoOps](/deploy-manage/monitor/autoops.md).
```

* From your deployment menu, view the [**Performance**](../../deploy-manage/monitor/access-performance-metrics-on-elastic-cloud.md) page's [memory pressure troubleshooting charts](/troubleshoot/monitoring/high-memory-pressure.md).
::::::

::::::{applies-item} { self:, eck: }
* (Recommend) Enable [AutoOps](/deploy-manage/monitor/autoops.md).
* Enable [{{es}} monitoring](/deploy-manage/monitor/stack-monitoring.md). When logs and metrics are enabled, monitoring information is visible on {{kib}}'s [Stack Monitoring](../../deploy-manage/monitor/monitoring-data/visualizing-monitoring-data.md) page. You can also enable the [JVM memory threshold alert](../../deploy-manage/monitor/monitoring-data/configure-stack-monitoring-alerts.md) to be notified about potential issues through email.
::::::

:::::::

## Reduce JVM memory pressure [reduce-jvm-memory-pressure]

This section contains some common suggestions for reducing JVM memory pressure.

### Common setup issues [reduce-jvm-memory-pressure-setup]

> This section contains some common suggestions for why JVM memory pressure can remain high in the background or respond non-linearly during performance issues.

Reviewer 2: Changed the intro to avoid repetition (multiple sentences starting with "this section contains some common suggestions"!)

> #### Disable swapping [reduce-jvm-memory-pressure-setup-swap]

Reviewer 2: ````suggestion
#### Disable swapping [reduce-jvm-memory-pressure-setup-swap]

```{applies_to}
deployment:
  ece: ga
  self: ga
  eck: ga
serverless: ga
```
````

If i'm not mistaken that should do the trick.

cc: @shainaraskas 

By the way I'd recommend to not abuse of this feature and use it only wherever it makes more sense.

For example in the JVM section you have a paragraph explaining that in ECH and ECE the platform shows you the memory pressure in the UI. That's perfectly fine, no need to mark or tag a paragraph if it's self-explanatory.

{{es}}'s JVM handles its own executables and can suffer severe performance degredation due to operating system swapping. We recommend [disabling swap](/deploy-manage/deploy/self-managed/setup-configuration-memory.md#bootstrap-memory_lock).

> Per guide, you can attempt to disable swap on the {{es}} level with [setting `bootstrap.memory_lock`](/deploy-manage/deploy/self-managed/setup-configuration-memory.md). In response, {{es}} will attempt to set `mlockall`; however this may fail. To check the setting and its outcome, poll the [node information API](https://www.elastic.co/docs/api/doc/elasticsearch/operation/operation-cat-nodes):

Reviewer: agree with edu that the duplicate content should be removed. this is a debt risk if this process changes in any way, so it's good to centralize the procedure in one place. if we want to duplicate it, we should use snippets, but I don't think that's necessary.

> Per guide, you can attempt to disable swap on the {{es}} level with [setting `bootstrap.memory_lock`](/deploy-manage/deploy/self-managed/setup-configuration-memory.md). In response, {{es}} will attempt to set `mlockall`; however this may fail. To check the setting and its outcome, poll the [node information API](https://www.elastic.co/docs/api/doc/elasticsearch/operation/operation-cat-nodes):

Author: The problem Support encounters & therefore checks progresses
1. users don't set `bootstrap.memory_lock` on ALL nodes
2. even if set `mlockall` doesn't trigger on ALL nodes (due to hardware profile / setup differences)
3. even if set+locked, the OS may ignore it anyway but swap is instantaneous so may false-positive at `0`

The intention isn't to repeat the original guide & I attempted to only TLDR say you set the setting which may/not take so check those and then check final impact by seeing if see swap. I can understand it feels duplicative though & am not sure how to improve as users not figuring how to check the original doc well enough is what got me here. Or perhaps the update should inject into original and then more simply link from here?

> Per guide, you can attempt to disable swap on the {{es}} level with [setting `bootstrap.memory_lock`](/deploy-manage/deploy/self-managed/setup-configuration-memory.md). In response, {{es}} will attempt to set `mlockall`; however this may fail. To check the setting and its outcome, poll the [node information API](https://www.elastic.co/docs/api/doc/elasticsearch/operation/operation-cat-nodes):

Author: Decided to update to just the swap portion we like & will consider adding to other doc.

```console
GET _nodes?filter_path=**.mlockall,**.memory_lock,nodes.*.name
```

For example, you can store this output into `nodes.json` and then using [third-party tool JQ](https://jqlang.github.io/jq/):

```bash
cat nodes.json | jq -rc '.nodes[]|.name as $n|{node:.name, memory_lock:.settings.bootstrap.memory_lock, mlockall:.process.mlockall}'
```

{{es}} recommends completely disabling swap on the operating system. This is because anything set {{es}}-level is best effort but swap can have severe impact on {{es}} performance. To check if any nodes are currently swapping, poll the [nodes stats API](https://www.elastic.co/docs/api/doc/elasticsearch/operation/operation-nodes-stats):

```console
GET _nodes/stats?filter_path=**.swap,nodes.*.name
```

> For example, you can store this output into `nodes_stats.json` and then using [third-party tool JQ](https://jqlang.github.io/jq/):

Reviewer 2: ```suggestion
For example, you can store this output into `nodes_stats.json` and then using [third-party tool JQ](https://jqlang.github.io/jq/) to process it:
```

The content feels a bit repetitive and the paragraphs about ES level swap and OS level swap feel unexpected or not presented. But anyway not a big deal.

```bash
cat nodes_stats.json | jq -rc '.nodes[]|{name:.name, swap_used:.os.swap.used_in_bytes}' | sort
```

> If nodes are found to be swapping after attempting to disable on the {{es}} level, you will need to escalate to [disabling swap operating system level](/deploy-manage/deploy/self-managed/setup-configuration-memory.md#disable-swap-files) to avoid performance impact.

Author: ```suggestion
If nodes are found to be swapping after attempting to disable on the {{es}} level, you need to escalate to [disabling swap on the operating system level](/deploy-manage/deploy/self-managed/setup-configuration-memory.md#disable-swap-files) to avoid performance impact.
```

> If nodes are found to be swapping after attempting to disable on the {{es}} level, you need to escalate to [disabling swap on the operating system level](/deploy-manage/deploy/self-managed/setup-configuration-memory.md#disable-swap-files) to avoid performance impact.

Reviewer 2: > If nodes are found to be swapping after attempting to disable on the {{es}} level, you need to escalate to [disabling swap on the operating system level](/deploy-manage/deploy/self-managed/setup-configuration-memory.md#disable-swap-files) to avoid performance impact.

That sentence is great, and gives a lot of value to this doc. If you see we don't tell the reader how to disable swap at OS level, we just link to the doc. IMO we should have done the same for ES level swapping configuration.

#### Enable compressed OOPs [reduce-jvm-memory-pressure-setup-oops]

> JVM performance strongly depends on having [Compressed OOPs](https://docs.oracle.com/javase/7/docs/technotes/guides/vm/performance-enhancements-7.html#compressedOop) enabled. The exact max heap size cut off depends on operating system, but usually caps near 30GB. To check if it is enabled, poll the [node information API](https://www.elastic.co/docs/api/doc/elasticsearch/operation/operation-cat-nodes):

Reviewer: noun of this is one word

`caps near` is a little casual

```suggestion
JVM performance strongly depends on having [Compressed OOPs](https://docs.oracle.com/javase/7/docs/technotes/guides/vm/performance-enhancements-7.html#compressedOop) enabled. The exact max heap size cutoff depends on operating system, but is typically around 30GB. To check if it is enabled, poll the [node information API](https://www.elastic.co/docs/api/doc/elasticsearch/operation/operation-cat-nodes):
```

```console
GET _nodes?filter_path=nodes.*.name,nodes.*.jvm.using_compressed_ordinary_object_pointers
```

For example, you can store this output into `nodes.json` and then using [third-party tool JQ](https://jqlang.github.io/jq/):

```bash
cat nodes.json | jq -rc '.nodes[]|{node:.name, compressed:.jvm.using_compressed_ordinary_object_pointers}'
```

> #### Max heap at less than half RAM [reduce-jvm-memory-pressure-setup-heap]

Reviewer 2: Or `Cofigure max heap size to less than...` if you want to keep the "max" term, as that would also be actionable :)

> By default, {{es}} manages the JVM heap size. If manually overridden, `Xms` and `Xmx` should be equal and not more than half of total operating system RAM per [Set the JVM heap size](elasticsearch://reference/elasticsearch/jvm-settings.md#set-jvm-heap-size). 

Reviewer 2: ```suggestion
By default, {{es}} manages the JVM heap size. If manually overridden, `Xms` and `Xmx` should be equal and not more than half of total operating system RAM. Refer to [Set the JVM heap size](elasticsearch://reference/elasticsearch/jvm-settings.md#set-jvm-heap-size) for detailed guidance and best practices.
```

Reviewer 2: I want to highlight the referenced doc.

To check these heap settings, poll the [node information API](https://www.elastic.co/docs/api/doc/elasticsearch/operation/operation-cat-nodes):

```console
GET _nodes?filter_path=nodes.*.name,nodes.*.jvm.mem
```

For example, you can store this output into `nodes.json` and then using [third-party tool JQ](https://jqlang.github.io/jq/):

```bash
cat nodes.json | jq -rc '.nodes[]|.name as $n|.jvm.mem|{name:$n, heap_min:.heap_init, heap_max:.heap_max}'
```

#### Reduce your shard count [reduce-jvm-memory-pressure-setup-shards]

> Every shard uses memory. In most cases, a small set of large shards uses fewer resources than many small shards. For tips on reducing your shard count, see [Size your shards](../../deploy-manage/production-guidance/optimize-performance/size-shards.md).

Author: ```suggestion
Every shard uses memory. Usually a small set of large shards uses fewer resources than many small shards. For tips on reducing your shard count, refer to [Size your shards](../../deploy-manage/production-guidance/optimize-performance/size-shards.md).
```

}
```

#### Prevent mapping explosion [reduce-jvm-memory-pressure-setup-mapping]

Defining too many fields or nesting fields too deeply can lead to [mapping explosions](/troubleshoot/elasticsearch/mapping-explosion.md) that use large amounts of memory. To prevent mapping explosions, use the [mapping limit settings](elasticsearch://reference/elasticsearch/index-settings/mapping-limit.md) to limit the number of field mappings.

> You might also consider setting {{kib}} [Advacned Settings](kibana://reference/advanced-settings.md) `data_views:fields_excluded_data_tiers` for faster performance. For example, to avoid searchable snapshots on cold and frozen [data tiers](/manage-data/lifecycle/data-tiers.md), you would set this to `data_cold,data_frozen`. This can help Discover fields to load faster as depicted in [Troubleshooting guide: Solving 6 common issues in Kibana Discover load](https://www.elastic.co/blog/troubleshooting-guide-common-issues-kibana-discover-load#2.-load-fields).

Reviewer 2: ```suggestion
You can also configure the {{kib}} [advanced setting](kibana://reference/advanced-settings.md) `data_views:fields_excluded_data_tiers` to improve performance by preventing {{kib}} from retrieving field data from specific data tiers. For example, to exclude cold and frozen tiers, typically used for searchable snapshots, set this value to `data_cold,data_frozen`. This can help Discover load fields faster, as described in [Troubleshooting guide: Solving 6 common issues in Kibana Discover load](https://www.elastic.co/blog/troubleshooting-guide-common-issues-kibana-discover-load#2.-load-fields).
```
